xc-vusb/xc-vusb.c

```diff
@@ -0,0 +1,2998 @@
+	if (err <= 0) {
+		eprintk("bind_evtchn_to_irqhandler failed device %p ret: %d\n",
+			vdev, err);
+		goto fail;
+	}
+	vdev->irq = err;
+
+	/* Allocate the shadow buffers */
+	vdev->shadows = kzalloc(sizeof(struct vusb_shadow)*SHADOW_ENTRIES,
+				GFP_KERNEL);
+	if (!vdev->shadows) {
+		xenbus_dev_fatal(dev, err, "allocate shadows failed");
+		err = -ENOMEM;
+		goto fail;
+	}
+
+	vdev->shadow_free_list = kzalloc(sizeof(u16)*SHADOW_ENTRIES,
+				GFP_KERNEL);
+	if (!vdev->shadow_free_list) {
+		xenbus_dev_fatal(dev, err, "allocate shadow list failed");
+		err = -ENOMEM;
+		goto fail;
+	}
+
+	for (i = 0; i < SHADOW_ENTRIES; i++) {
+		vdev->shadows[i].req.id = i;
+		vdev->shadows[i].in_use = 1;
+		vusb_put_shadow(vdev, &vdev->shadows[i]);
+	}
+
+	return 0;
+fail:
+	vusb_usbif_free(vdev);
+	return err;
+}
+
+/* Common code used when first setting up, and when resuming. */
+static int
+vusb_talk_to_usbback(struct vusb_device *vdev)
+{
+	struct xenbus_device *dev = vdev->xendev;
+	const char *message = NULL;
+	struct xenbus_transaction xbt;
+	int err;
+
+	/* Create shared ring, alloc event channel. */
+	err = vusb_setup_usbfront(vdev);
+	if (err)
+		goto out;
+
+again:
+	err = xenbus_transaction_start(&xbt);
+	if (err) {
+		xenbus_dev_fatal(dev, err, "starting transaction");
+		goto free_usbif;
+	}
+
+	err = xenbus_printf(xbt, dev->nodename,
+	 		"ring-ref", "%u", vdev->ring_ref);
+	if (err) {
+		message = "writing ring-ref";
+		goto abort_transaction;
+	}
+
+	err = xenbus_printf(xbt, dev->nodename,
+			"event-channel", "%u", vdev->evtchn);
+	if (err) {
+		message = "writing event-channel";
+		goto abort_transaction;
+	}
+
+	err = xenbus_printf(xbt, dev->nodename, "version", "%d",
+			VUSB_INTERFACE_VERSION);
+	if (err) {
+		message = "writing protocol";
+		goto abort_transaction;
+	}
+
+	err = xenbus_transaction_end(xbt, 0);
+	if (err) {
+		if (err == -EAGAIN)
+			goto again;
+		xenbus_dev_fatal(dev, err, "completing transaction");
+		goto free_usbif;
+	}
+
+	/* Started out in the initialising state, go to initialised */
+	xenbus_switch_state(dev, XenbusStateInitialised);
+
+	return 0;
+
+ abort_transaction:
+	xenbus_transaction_end(xbt, 1);
+	if (message)
+		xenbus_dev_fatal(dev, err, "%s", message);
+ free_usbif:
+	vusb_usbif_free(vdev);
+ out:
+	return err;
+}
+
+static int
+vusb_create_device(struct vusb_vhcd *vhcd, struct xenbus_device *dev, u16 id)
+{
+	int ret = 0;
+	struct vusb_rh_port *vport;
+	struct vusb_device *vdev;
+	unsigned long flags;
+
+	/* Find a port we can use. */
+	spin_lock_irqsave(&vhcd->lock, flags);
+	vport = vusb_get_vport(vhcd, id);
+	spin_unlock_irqrestore(&vhcd->lock, flags);
+	if (!vport)
+		return -ENODEV;
+
+	vdev = vusb_vdev_by_port(vhcd, vport->port);
+
+	spin_lock_init(&vdev->lock);
+	INIT_LIST_HEAD(&vdev->pending_list);
+	INIT_LIST_HEAD(&vdev->release_list);
+	INIT_LIST_HEAD(&vdev->finish_list);
+	INIT_WORK(&vdev->work, vusb_device_work_handler);
+	init_waitqueue_head(&vdev->wait_queue);
+
+	/* Strap our vUSB 
```

Comment:
Yea that is code that is incomplete. I am not sure that unloading the vHCD driver is ever going to be safe. I was leaning towards a udev rule to prevent it being unloaded. There are just too many synchronization issues with trying to unload it with active xenbus devices.